src/laddie/laddie_thickness.f90

```diff
@@ -0,0 +1,198 @@
+        laddie%detr( vi) = - MIN(laddie%entr( vi),0.0_dp)
+
+        ! Get actual dHdt
+        dHdt = -laddie%divQH( vi) + laddie%melt( vi) + laddie%entr( vi) + laddie%entr_dmin( vi)
+
+        ! Get actual H_n
+        npx%H( vi) = laddie%now%H( vi) + dHdt * dt
+
+      END IF !(laddie%mask_a( vi)) THEN
+    END DO !vi = mesh%vi, mesh%v2
+
+    ! Map new values of H to b grid and c grid
+    CALL map_H_a_b( mesh, laddie, npx%H, npx%H_b)
+    CALL map_H_a_c( mesh, laddie, npx%H, npx%H_c)
+
+    CALL check_for_NaN_dp_1D( npx%H, 'H_lad')
+
+    ! Finalise routine path
+    CALL finalise_routine( routine_name)
+
+  END SUBROUTINE compute_H_npx
+
+  SUBROUTINE compute_divQH( mesh, laddie, npx)
+    ! Divergence of layer thickness based on first order upwind scheme
+
+    ! In/output variables:
+    TYPE(type_mesh),                        INTENT(IN)    :: mesh
+    TYPE(type_laddie_model),                INTENT(INOUT) :: laddie
+    TYPE(type_laddie_timestep),             INTENT(IN)    :: npx
+
+    ! Local variables:
+    CHARACTER(LEN=256), PARAMETER                         :: routine_name = 'compute_divQH'
+    REAL(dp), DIMENSION(mesh%nE)                          :: U_c_tot, V_c_tot
+    REAL(dp), DIMENSION(mesh%nV)                          :: H_tot
+    INTEGER                                               :: vi, ci, vj, ei
+    REAL(dp)                                              :: D_x, D_y, D, u_perp
+    LOGICAL, DIMENSION(mesh%nV)                           :: mask_gr_a_tot, mask_oc_a_tot
+
+    ! Add routine to path
+    CALL init_routine( routine_name)
+
+    ! Calculate vertically averaged ice velocities on the edges
+    CALL gather_to_all_dp_1D( npx%U_c, U_c_tot)
+    CALL gather_to_all_dp_1D( npx%V_c, V_c_tot)
+    CALL gather_to_all_dp_1D( npx%H, H_tot)
+    CALL gather_to_all_logical_1D( laddie%mask_gr_a, mask_gr_a_tot)
+    CALL gather_to_all_logical_1D( laddie%mask_oc_a, mask_oc_a_tot)
+
+    ! Initialise with zeros
+    laddie%divQH = 0.0_dp
+
+    ! == Loop over vertices ==
+    ! =========================
+
+    DO vi = mesh%vi1, mesh%vi2
+
+      IF (laddie%mask_a( vi)) THEN
+        ! Initialise
+
+        ! Loop over all connections of vertex vi
+        DO ci = 1, mesh%nC( vi)
+
+          ! Connection ci from vertex vi leads through edge ei to vertex vj
+          vj = mesh%C(  vi,ci)
+
+          IF (vj == 0) CYCLE
+
+          ! Skip connection if neighbour is grounded. No flux across grounding line
+          ! Can be made more flexible when accounting for partial cells (PMP instead of FCMP)
+          IF (mask_gr_a_tot( vj)) CYCLE
+
+          ! Get edge
+          ei = mesh%VE( vi,ci)
+
+          ! Calculate vertically averaged ice velocity component perpendicular to this shared Voronoi cell boundary section
+          D_x = mesh%V( vj,1) - mesh%V( vi,1)
+          D_y = mesh%V( vj,2) - mesh%V( vi,2)
+          D   = SQRT( D_x**2 + D_y**2)
+          u_perp = U_c_tot( ei) * D_x/D + V_c_tot( ei) * D_y/D
```

Comment:
Yes, I had this saved for #99 as it will significantly reduce computation time. Wasn't sure yet whether I'd need each of these, but am now. So I'll move this to mesh_secondary